code/datums/components/rust.dm

```diff
@@ -0,0 +1,87 @@
+/**
+ * Adding this component to an Atom will have it automatically render an overlay.
+ * The overlay can be specified in new as the first and only parameter; it defaults to "rust" if not included.
+ */
+/datum/component/rust
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+	/// Internal variable to store the rust overlay we are using to avoid regenerating it every overlay update call
+	var/mutable_appearance/rust_overlay
+
+/turf/closed/wall/rust/New()
+	var/atom/wall_new = new /turf/closed/wall(src)
+	wall_new._AddComponent(list(/datum/component/rust))
+
+/turf/closed/wall/r_wall/rust/New()
+	var/atom/wall_new = new /turf/closed/wall/r_wall(src)
+	wall_new._AddComponent(list(/datum/component/rust))
+
+/turf/open/floor/plating/rust/New()
+	var/atom/wall_new = new /turf/open/floor/plating(src)
+	wall_new._AddComponent(list(/datum/component/rust))
+
+/datum/component/rust/Initialize(rust_iconstate = "rust")
+	. = ..()
+	if(!isatom(parent))
+		stack_trace("Attempted to intialize a Rust Component for non-Atom [parent]")
```

Comment:
COMPONENT_INCOMPATIBLE comes with its own stack_trace messages IIRC, this is not necessary.